tests/lightning_qubit/test_measurements_class.py

```diff
@@ -379,6 +391,155 @@ def test_sparse_Pauli_words(self, ham_terms, expected, tol, lightning_sv):
         assert np.allclose(result, expected, tol)
 
 
+class TestMeasurements:
+    """Tests all measurements"""
+
+    @staticmethod
+    def calculate_reference(tape, lightning_sv):
+        use_default = True
+        new_meas = []
+        for m in tape.measurements:
+            # not supported by DefaultQubit
+            if isinstance(m, VarianceMP) and isinstance(
+                m.obs, (qml.Hamiltonian, qml.SparseHamiltonian)
+            ):
+                use_default = False
+                new_meas.append(m.__class__(qml.Hermitian(qml.matrix(m.obs), wires=m.obs.wires)))
+                continue
+            new_meas.append(m)
+        if use_default:
+            dev = DefaultQubit(max_workers=1)
+            program, _ = dev.preprocess()
+            tapes, transf_fn = program([tape])
+            results = dev.execute(tapes)
+            return transf_fn(results)
+
+        tape = qml.tape.QuantumScript(tape.operations, new_meas)
+        statevector = lightning_sv(tape.num_wires)
+        statevector = statevector.get_final_state(tape)
+        m = LightningMeasurements(statevector)
+        return m.measure_final_state(tape)
+
+    @pytest.mark.parametrize("measurement", [qml.expval, qml.probs, qml.var])
+    @pytest.mark.parametrize(
+        "observable",
+        (
+            [0],
+            [1, 2],
+            qml.PauliX(0),
+            qml.PauliY(1),
+            qml.PauliZ(2),
+            qml.sum(qml.PauliX(0), qml.PauliY(0)),
+            qml.prod(qml.PauliX(0), qml.PauliY(1)),
+            qml.s_prod(2.0, qml.PauliX(0)),
+            qml.Hermitian(get_hermitian_matrix(2**2), wires=[0, 1]),
+            qml.Hermitian(get_hermitian_matrix(2**2), wires=[2, 3]),
+            qml.Hamiltonian(
+                [1.0, 2.0, 3.0], [qml.PauliX(0), qml.PauliY(1), qml.PauliZ(2) @ qml.PauliZ(3)]
+            ),
+            qml.SparseHamiltonian(get_sparse_hermitian_matrix(2**4), wires=range(4)),
+        ),
+    )
+    def test_single_return_value(self, measurement, observable, lightning_sv, tol):
+        if measurement is qml.probs and isinstance(
+            observable,
+            (qml.ops.Sum, qml.ops.SProd, qml.ops.Prod, qml.Hamiltonian, qml.SparseHamiltonian),
+        ):
+            return
+        if measurement is not qml.probs and isinstance(observable, list):
+            return
+        n_qubits = 4
+        n_layers = 1
+        np.random.seed(0)
+        weights = np.random.rand(n_layers, n_qubits, 3)
+        ops = [qml.Hadamard(i) for i in range(n_qubits)]
+        ops += [qml.StronglyEntanglingLayers(weights, wires=range(n_qubits))]
+        measurements = (
+            [measurement(wires=observable)]
+            if isinstance(observable, list)
+            else [measurement(op=observable)]
+        )
+        tape = qml.tape.QuantumScript(ops, measurements)
+
+        expected = self.calculate_reference(tape, lightning_sv)
+        statevector = lightning_sv(n_qubits)
```

Comment:
This seems problematic for Variance with `Hamiltonian` and `SparseHamiltonian` as both the expected and actual result will be computed by `LightningMeasurements`.

Reply:
Yes, this is because I hit a `NotImplementedError` error in DQ. `calculate_reference` therefore uses `qml.Hermitian`, which is validated independently` for validation.